File: Assets/Bundles/Server/Resources/TS/Services/Match/ShopkeeperService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ObjectUtil = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local NetworkBridge = require("Shared/TS/NetworkBridge").NetworkBridge
local Task = require("Shared/TS/Util/Task").Task
local MathUtil = require("Shared/TS/Util/MathUtil").MathUtil
-- * Generator deny region size.
local DENY_REGION_SIZE = Vector3.new(2, 3, 2)
local ShopkeeperService
do
	ShopkeeperService = setmetatable({}, {
		__tostring = function()
			return "ShopkeeperService"
		end,
	})
	ShopkeeperService.__index = ShopkeeperService
	function ShopkeeperService.new(...)
		local self = setmetatable({}, ShopkeeperService)
		return self:constructor(...) or self
	end
	function ShopkeeperService:constructor(mapService, denyRegionService)
		self.mapService = mapService
		self.denyRegionService = denyRegionService
		self.shopKeeperPrefab = AssetBridge:LoadAsset("Shared/Resources/Entity/HumanEntity/HumanEntity.prefab")
	end
	function ShopkeeperService:OnStart()
		Task:Spawn(function()
			-- Wait map and match started before creating generators.
			self.loadedMap = self.mapService:WaitForMapLoaded()
			ServerSignals.MatchStart:connect(function()
				return self:CreateShopKeepers()
			end)
		end)
	end
	function ShopkeeperService:CreateShopKeepers()
		-- Shop shopkeepers.
		local shops = self.loadedMap:GetAllShopkeepers()
		local _exp = ObjectUtil.values(shops)
		local _arg0 = function(mapPosition)
			local keeper = GameObjectBridge:InstantiateAt(self.shopKeeperPrefab, mapPosition.Position, mapPosition.Rotation)
			NetworkBridge:Spawn(keeper, "item_shop_shopkeeper")
			-- Create deny region around shopkeeper.
			self.denyRegionService:CreateDenyRegion(MathUtil:FloorVec(mapPosition.Position), DENY_REGION_SIZE)
		end
		for _k, _v in _exp do
			_arg0(_v, _k - 1, _exp)
		end
		-- Team upgrade shopkeepers.
		local upgrades = self.loadedMap:GetAllTeamUpgrades()
		local _exp_1 = ObjectUtil.values(upgrades)
		local _arg0_1 = function(mapPosition)
			local keeper = GameObjectBridge:InstantiateAt(self.shopKeeperPrefab, mapPosition.Position, mapPosition.Rotation)
			NetworkBridge:Spawn(keeper, "team_upgrades_shopkeeper")
			-- Create deny region around shopkeeper.
			self.denyRegionService:CreateDenyRegion(MathUtil:FloorVec(mapPosition.Position), DENY_REGION_SIZE)
		end
		for _k, _v in _exp_1 do
			_arg0_1(_v, _k - 1, _exp_1)
		end
	end
end
-- (Flamework) ShopkeeperService metadata
Reflect.defineMetadata(ShopkeeperService, "identifier", "Bundles/Server/Services/Match/ShopkeeperService@ShopkeeperService")
Reflect.defineMetadata(ShopkeeperService, "flamework:parameters", { "Bundles/Server/Services/Match/Map/MapService@MapService", "Bundles/Server/Services/Global/Block/DenyRegionService@DenyRegionService" })
Reflect.defineMetadata(ShopkeeperService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ShopkeeperService, "$:flamework@Service", Service, { {} })
return {
	ShopkeeperService = ShopkeeperService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Match/ShopkeeperService.ts
import { CoreServerSignals } from "@Easy/Core/Server/CoreServerSignals";
import { DenyRegionService } from "@Easy/Core/Server/Services/Block/DenyRegionService";
import { EntityService } from "@Easy/Core/Server/Services/Entity/EntityService";
import { TeamService } from "@Easy/Core/Server/Services/Team/TeamService";
import { EntityPrefabType } from "@Easy/Core/Shared/Entity/EntityPrefabType";
import { ColorUtil } from "@Easy/Core/Shared/Util/ColorUtil";
import { MathUtil } from "@Easy/Core/Shared/Util/MathUtil";
import { Theme } from "@Easy/Core/Shared/Util/Theme";
import { OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { Network } from "Shared/Network";
import { MapService } from "./Map/MapService";

/** Generator deny region size. */
const DENY_REGION_SIZE = new Vector3(2, 3, 2);

@Service({})
export class ShopkeeperService implements OnStart {
	/** Shopkeeper prefab. */
	private shopKeeperPrefab: Object;
	private itemShopEntityIds: number[] = [];
	private upgradeShopEntityIds: number[] = [];

	constructor(
		private readonly mapService: MapService,
		private readonly denyRegionService: DenyRegionService,
		private readonly teamService: TeamService,
		private readonly entityService: EntityService,
	) {
		this.shopKeeperPrefab = AssetBridge.Instance.LoadAsset(
			"@Easy/Core/Shared/Resources/Entity/HumanEntity/HumanEntity.prefab",
		);

		ServerSignals.MapLoad.Connect(() => {
			this.CreateShopKeepers();
		});
	}

	OnStart(): void {
		CoreServerSignals.PlayerJoin.Connect((event) => {
			Network.ServerToClient.ItemShop.AddNPCs.Server.FireClient(event.player.clientId, this.itemShopEntityIds);
			Network.ServerToClient.TeamUpgradeShop.AddNPCs.Server.FireClient(
				event.player.clientId,
				this.upgradeShopEntityIds,
			);
		});
	}

	/** Creates shop shopkeepers and team upgrade shopkeepers. */
	private CreateShopKeepers(): void {
		const loadedMap = this.mapService.GetLoadedMap();
		if (!loadedMap) {
			Debug.LogError("Loaded map was not ready when spawning shopkeepers.");
			return;
		}

		this.itemShopEntityIds.clear();
		this.upgradeShopEntityIds.clear();
		for (let team of this.teamService.GetTeams()) {
			// Item Shop
			const itemShopWorldPos = loadedMap.GetWorldPosition(team.id + "_upgrade_shop");
			const itemShopEntity = this.entityService.SpawnEntityForPlayer(
				undefined,
				EntityPrefabType.HUMAN,
				itemShopWorldPos.Position,
			);
			itemShopEntity.entityDriver.SetLookVector(
				itemShopWorldPos.Rotation.mul(itemShopEntity.entityDriver.transform.forward),
			);
			itemShopEntity.SetDisplayName(ColorUtil.ColoredText(Theme.Yellow, "Item Shop"));
			itemShopEntity.GrantImmunity(math.huge);
			this.itemShopEntityIds.push(itemShopEntity.id);

			this.denyRegionService.CreateDenyRegion(MathUtil.FloorVec(itemShopWorldPos.Position), DENY_REGION_SIZE);

			// Team Upgrades
			const teamUpgradeWorldPos = loadedMap.GetWorldPosition(team.id + "_item_shop");
			const upgradeShopEntity = this.entityService.SpawnEntityForPlayer(
				undefined,
				EntityPrefabType.HUMAN,
				teamUpgradeWorldPos.Position,
			);
			upgradeShopEntity.entityDriver.SetLookVector(
				teamUpgradeWorldPos.Rotation.mul(upgradeShopEntity.entityDriver.transform.forward),
			);
			upgradeShopEntity.SetDisplayName(ColorUtil.ColoredText(Theme.Yellow, "Team Upgrades"));
			upgradeShopEntity.GrantImmunity(math.huge);
			this.upgradeShopEntityIds.push(upgradeShopEntity.id);

			this.denyRegionService.CreateDenyRegion(MathUtil.FloorVec(teamUpgradeWorldPos.Position), DENY_REGION_SIZE);
		}

		Network.ServerToClient.ItemShop.AddNPCs.Server.FireAllClients(this.itemShopEntityIds);
		Network.ServerToClient.TeamUpgradeShop.AddNPCs.Server.FireAllClients(this.upgradeShopEntityIds);
	}
}
